src/DynamicData/Binding/IObservableListEx.cs

```diff
@@ -0,0 +1,197 @@
+        }
+
+        /// <summary>
+        /// Binds the results to the specified <see cref="IObservableList{T}"/>. Unlike
+        /// binding to a <see cref="ReadOnlyObservableCollection{T}"/> which loses the
+        /// ability to refresh items, binding to an <see cref="IObservableList{T}"/>.
+        /// allows for refresh changes to be preserved and keeps the list read-only.
+        /// </summary>
+        /// <typeparam name="TObject">The type of the object.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="source">The source.</param>
+        /// <param name="destination">The destination.</param>
+        /// <returns>The <paramref name="source"/> changeset for continued chaining.</returns>
+        /// <exception cref="System.ArgumentNullException">source</exception>
+        public static IObservable<ISortedChangeSet<TObject, TKey>> BindToObservableList<TObject, TKey>(
+            this IObservable<ISortedChangeSet<TObject, TKey>> source,
+            out IObservableList<TObject> observableList)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            // Load our source list with the change set.
+            // Each changeset we need to convert to remove the key.
+            var sourceList = new SourceList<TObject>();
+
+            // Output our readonly observable list, preventing the sourcelist from being editted from anywhere else.
+            observableList = sourceList;
+
+            // Return a observable that will connect to the source so we can properly dispose when the pipeline ends.
+            return Observable.Create<ISortedChangeSet<TObject, TKey>>(observer =>
+            {
+                return source
+                    .Do(changes => sourceList.Edit(editor => editor.Clone(changes.Convert(editor))))
+                    .Finally(() => sourceList.Dispose())
+                    .SubscribeSafe(observer);
+            });
+        }
+
+        /// <summary>
+        /// Converts a <see cref="IChangeSet{TObject, TKey}"/> to <see cref="IChangeSet{TObject}"/>
+        /// which allows for binding a cache to a list.
+        /// </summary>
+        /// <typeparam name="TObject">The type of the object.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="changeSetWithKey">The source change set</param>
+        /// <param name="list">
+        /// An optional list, if provided it allows the refresh from a key based cache to find the index for the resulting list based refresh.
+        /// If not provided a refresh will dropdown to a replace which may ultimately result in a remove+add change downstream.
+        /// </param>
+        /// <returns>The downcasted <see cref="IChangeSet{TObject}"/></returns>
+        private static IChangeSet<TObject> Convert<TObject, TKey>(this IChangeSet<TObject, TKey> changeSetWithKey, IExtendedList<TObject> list = null)
```

Comment:
Have you considered using, or adapting the `RemoveKey` operator.

This serves a similar purpose to what you are doing here.  Internally remove key uses `RemoveKeyEnumerator<TObject, TKey>` which is close to what you want. The only obvious difference is that an update is treated as an add / remove rather than a replace. Also we would need to ensure it handles the refresh.

Reply:
Thanks for this tip, we were trying to find something like this in the codebase but just couldn't seem to find it.

We adapted to use `RemoveKeyEnumerator<TObject, TKey>`. Our unit tests for refresh started to fail. Additionally, we wrote a unit test for move, which also started to fail. So we improved `RemoveKeyEnumerator<TObject, TKey>` to handle these two change reasons. We also improved it to pass through the index since this would be more performant when cloning out to the list.